README.md
# BEdita, a back-end API


[![Build Status](https://github.com/bedita/bedita/workflows/Run%20tests/badge.svg)](https://github.com/bedita/bedita/actions?query=workflow%3A%22Run+tests%22)
[![Publish plugins](https://github.com/bedita/bedita/workflows/Publish%20components/badge.svg)](https://github.com/bedita/bedita/actions?query=workflow%3A%22Publish+components%22)
[![Code Coverage](https://codecov.io/gh/bedita/bedita/branch/4-cactus/graph/badge.svg)](https://codecov.io/gh/bedita/bedita/branch/4-cactus)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/bedita/bedita/badges/quality-score.png?b=4-cactus)](https://scrutinizer-ci.com/g/bedita/bedita/?branch=4-cactus)

BEdita 4 is a ready to use back-end API to handle the data of your mobile, IoT, web and desktop applications.
It's also an extensible framework to build your custom back-end API via plugins.

It provides a complete content management solution with:

* a _headless_ HTTP server application with a complete REST API to model, create, modify and retrieve data
* a default admin web application **[BEdita Manager](https://github.com/bedita/manager)**

BEdita 4 is built with [CakePHP 3](http://cakephp.org) and uses relational DBMS like [MySQL](http://www.mysql.com),
[Postgres](https://www.postgresql.org) or [SQLite](http://sqlite.com) in conjunction with (optional) NoSQL systems like [Redis](http://redis.io/), [Elastic Search](https://www.elastic.co/) or time series databases to boost performance and scale up to Big Data scenarios.

[JSON-API](http://jsonapi.org) is the primary exchange data format.

[GrapQL](http://graphql.org) initial support available via dedicated plugin.

The easiest and quickest way to try out BEdita4 is via [Docker](https://www.docker.com), [read instructions below](#docker).

## Prerequisites

* PHP 7.2, 7.3 or 7.4 with extensions: `json`, `mbstring`, `fileinfo`, `intl`, `pdo`, `simplexml`
* MySQL 5.7 (recommended) or MySQL 8.0, Postgres 9.6/10/11/12, MariaDB 10.2/10.3/10.4 or SQLite 3
* [Composer](https://getcomposer.org/doc/00-intro.md#installation-linux-unix-osx)

## Install

For a detailed setup overview see [Setup Documentation](https://docs.bedita.net/en/latest/setup.html)

Quick setup in three steps.

01. Create project via composer

 ```bash
  composer create-project bedita/bedita
 ```

 If you are using a **.zip** or **.tar.gz** release file you just need to unpack it and then run ``composer install``.

02. Create an empty database either MySQL or PostgresSQL. Do nothing for SQLite.

02. Change to the newly created folder and run this script to initialize the database and create first admin user:

```bash
 bin/cake bedita setup
```

See [Web Server setup](https://docs.bedita.net/en/latest/setup.html#web-server)
to configure a virtualhost in your webserver.
To see first [`/home` endpoint](https://docs.bedita.net/en/latest/endpoints/home.html) response you may point to `http://your-vhost/home`

Curl example:

```bash
 curl -H Accept:application/json http://your-vhost/home
```

Otherwise, only for development or test setups, you can take advantage of PHP builtin server with this simple command:

```bash
 bin/cake server
```

and see `/home` endpoint response pointing to `http://localhost:8765/home` like this:

```bash
 curl -H Accept:application/json http://localhost:8765/home
```

For an explanation on `Accept` headers usage [read here](https://docs.bedita.net/en/latest/endpoints/intro.html#headers).

## Docker

See [Docker setup documentation](https://docs.bedita.net/en/latest/setup.html#setup-docker) for a more detailed overview.

### Pull official image

Get latest offical image build from Docker Hub

```bash
 docker pull bedita/bedita:latest
```

You may also use `:4-cactus` tag instead of `:latest`, they are currently synonyms.

### Build image

If you want to build an image from local sources you can do it like this from BEdita root folder:

```bash
 docker build -t bedita4-local .
```

You may of course choose whatever name you like for the generated image instead of `bedita4-local`.

### Run

Run a Docker image setting an initial API KEY and admin username and password, like this:

```bash
 docker run -p 8090:80 --env BEDITA_API_KEY=1029384756 \
    --env BEDITA_ADMIN_USR=admin --env BEDITA_ADMIN_PWD=admin \
    bedita/bedita:latest
```

This will launch a BEdita4 instance using `SQLite` as its storage backend. It should become available at http://localhost:8090/home almost instantly.

Replace `bedita/bedita:latest` with `bedita4-local` (or other chosen name) to lanch a local built image.

## Documentation

Developer documentation can be found [here](https://docs.bedita.net)

## Licensing

BEdita is released under [LGPL](/bedita/bedita/blob/master/LICENSE.LGPL), Lesser General Public License v3.


CHANGES
Version 3.0.1 - betula
 User-visible changes:
 	* replace publishing with publication
 	* fixed wrong behavior on select country in addressbook (added jquery.selectboxes plugin)
 	* fix video thumbnails 
 	* fix <select> in newsletter subscription method (single opt-in, double opt-in)
 
 Frontend changes:
 	* fix including nickname.tpl(.ctp) view if exists and use views/pages instead of views/sections directory
 	* fix rss urls/links
 	* captcha generation bug solved (waiting for CakePHP to solve the srand bug ticket #140)
 	* fix canonicalPath
 
 Developer-visibile changes:
 	* replacing deprecated videoThumbHeight with media.video.thumbHeight
 	* small shell fixes
 	* multiple recipients for mailSupport ("to"=>'addr1@example.com, addr2@example.com')
 	* fix build filter error with keyword rel_object_id
 
  

Version 3.0 - betula

 User-visible changes:
 	* #196 - print layout, standard BEdita report. Introduced print option for BEdita objects 
 	* #81 - upload files directly in download relation tab
 	* fix multimedia items drag&drop and sort in IE7
 	* complete UI english/italian localization
 	* fix upload from URL
 	* fix video thumbnails in newsletter messages
 	* automatic remove of time in flowplayer when player width is too small
 	* #221 - fixed number of subscribed in newsletter main page
 	* #98 - improved newsletter module interface
 	* addressbook: company name for a person
 	* categories in Documents module
 	* vimeo video: get thumbnail_medium instead of large
 	* newsletter email (html) send in correct format (with -alt-- and boundary code between plain text and html type)
	* added link (continue) at the end of contents inserted in a newsletter message (if use template associated to a publication)
	* added "reply to" field in template form
	* default card status ON
	* many fix in addressbook module
	* fix ajax bug in operations on list subscribers
	* newsletter: added [$abstract] and [$subscriber] placeholders. Truncate options on [$description] and [$abstract]
	* bugfix - click on box "translations" back to translations module
	* #228 ui:preserve list filter on bulk operations and single delete
	* remove notes from mail group details
	* fix upload file in Windows
	
 
 Frontend changes:
 	* beEmbedFlash helper include swfobject automaticaly from BEdita backend using BEDITA_CORE_PATH
 	* #222 - staging toolbar for every frontend
 	* restyled site.example.com (html and css) 
 	* comments: skip capthca and use users data for users logged in frontend
 	* custom properties array reformatted with label as key
 	* #132 many fix on sitemap (xml and html)
 	* #211 canonicalPath for helpers, and pages/templates
 	* fix urls -> use .canonicalPath
 	
 
 Developer-visibile changes:
 	* php 5.3 support
	* update cakephp to 1.2.5
	* update to swfobject 2.2
	* update to uploadify 2.1.0
	* update to flowplayer to 3.1.4
	* update webroot/index.php and core.php.sample in backend and frontends
	* #196 - print layout, create a structure to build own frontend print layouts
	* beEmbedFlash helper now support flowplayer playlists
	* beEmbedFlash embed Flowplayer useing flowplayer.min.js instead of swfobject for IE7 problems
	* introduced Alias model for create nickname aliases
	* shell: dbadmin orphans, search and remove orphan objects (not in tree)
	* fix unit tests
	* #219 - fix delete of subsections when remove a section/publication
	* Tree unit test
	* #194 - delete wrong store procedure to clean tree. Cleaning tree on deleting object done through php code
	* #220 - fixed error on bulk actions
	* fixed clone card (added specific javascript)
	* change enum options in mail_status field
	* i18n update
	* #224 - vcard import, shell script -- merge
	* add AppHelper base class in backend and include it in frontends
	* refactor for apidoc
	* fix 'fixed' handling
	* update comments on bedita.ini, bedita.sys, bedita.cfg
	* refactoring of media_providers config array in bedita.ini
	* deploy script and release.cfg fix
	* workaround to avoid phpThum bug when use imagemagick and upscale=false
	* added duration in blip and vimeo videos
	* some changes on helpers video to use oEmbed format
	* #206 - partial nickname alias handling -- load frontend and object save
	* beEmbedFlash helper: option for embed flash object directly with html code (no javascript used)
	* patched cakePHP EmailComponent on line 493 to create boundary when email sent in text and html (both mode) 
	* shell newsletter: method to add cards to mail_group
	* bugfix - delete multimedia item with no file associated
	* added $fullBaseUrl attribute of AppController class. Used in forward methods
	* newsletter-send: managed BeditaException in sendQueuedJobs and create jobs only for "ON" card
	* shell script: dbadmin bulk import from csv
	* shell newsletter: mail group lists merge

bedita-app/config/bedita.ini.php
<?php
/*-----8<--------------------------------------------------------------------
 * 
 * BEdita - a semantic content management framework
 * 
 * Copyright 2008 ChannelWeb Srl, Chialab Srl
 * 
 * This file is part of BEdita: you can redistribute it and/or modify
 * it under the terms of the Affero GNU General Public License as published 
 * by the Free Software Foundation, either version 3 of the License, or 
 * (at your option) any later version.
 * BEdita is distributed WITHOUT ANY WARRANTY; without even the implied 
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Affero GNU General Public License for more details.
 * You should have received a copy of the Affero GNU General Public License 
 * version 3 along with BEdita (see LICENSE.AGPL).
 * If not, see <http://gnu.org/licenses/agpl-3.0.html>.
 * 
 *------------------------------------------------------------------->8-----
 */

/**
 * bedita.ini.php - settings, constants, variables for BEdita local installation
 * 
 * @link			http://www.bedita.com
 * @version			$Revision$
 * @modifiedby 		$LastChangedBy$
 * @lastmodified	$LastChangedDate$
 * 
 * $Id$
 */

$config = array() ;
 
$config["projectName"] = ""; // override in bedita.cfg

$config["userVersion"] = "BEdita 3.0.1"; // don't override

$config["majorVersion"] = "3.0.1"; // don't override -- admin/system

$config["helpBaseUrl"] = "http://docs.bedita.com/behelp/v30"; // don't override -- 

// Multimedia - image file substituting missing content (as now used in BeThumb helper)
$config['imgMissingFile'] = "/img/iconMissingImage_130x85.gif" ;

/**
 ** ******************************************
 **  Content and UI Elements defaults
 ** ******************************************
 */

// User Interface default language [see also 'multilang' below]
$config['Config']['language'] = "eng"; // or "ita", "eng", "spa", "por"

// Set 'multilang' true for user choice [also set 'multilang' true if $config['Config']['language'] is set]
$config['multilang'] = true;
$config['defaultLang'] = "eng"; // default fallback


/* Dates - default presentation format [syntax used by strftime php function]
 * It is used for datepicker calendar. Format has to contain %d,%m,%Y tags
 */
$config['datePattern'] 	= "%m-%d-%Y" ;

/* Date patterns for different locales
 * It is used for datepicker calendar. Format has to contain %d,%m,%Y tags
 */
$config['datePatternLocale'] = array(
	"ita"	=> "%d/%m/%Y"
);

// Default date time pattern
$config['dateTimePattern'] 	= "%m-%d-%Y  %H:%M:%S";  //

$config['dateTimePatternLocale'] = array(
	"ita"	=> "%d/%m/%Y %H:%M:%S" 
);


// Default status of new objects
$config['defaultStatus'] = "draft" ;

// TinyMCE Rich Text Editor for long_text ['true' to enable]
$config['mce'] = true;

// Application messages - temporary messages duration
$config['msgPause'] = 3000;		// milliseconds

// bedita user for unit test
$config['unitTestUserId'] = 1;

// activity timeout used in Home/"connected user" time duration in minutes
$config['activityTimeout'] = 20;

/**
 ** ******************************************
 **  Image, Video and Audio defaults
 ** ******************************************
 */
$config['media']['image']['thumbWidth']  = 130;      // px thumb width
$config['media']['image']['thumbHeight'] = 85;       // px thumb height
$config['media']['image']['thumbMode']   = "crop";   // crop, fill, croponly, stretch - thumb mode
$config['media']['image']['thumbFill']   = "FFFFFF"; // hex - fill color when thumb mode is fill
$config['media']['image']['thumbCrop']   = "C";      // string, crop mode when thumb mode is crop/croponly 'C', 'T', 'B', 'L', 'R', 'TL', 'TR', 'BL', 'BR'
$config['media']['image']['thumbQ']      = 75;       // int, JPEG thumbnail image quality [1-100]
$config['media']['image']['thumbUpscale']= false;     // bool, allow thumbnail upscale
$config['media']['image']['imagemagick'] = "";       // string, path to image_magick executable
$config['media']['image']['over']        = "";       // string, path to overlay image
$config['media']['image']['wmi']['f']    = "";       // string, path to watermark image file
$config['media']['image']['wmi']['a']    = "C";      // string, wm alignment B=bottom, T=top, L=left, R=right, C=centre, *=tile, 2 letters ie TL, or absolute position in px
$config['media']['image']['wmi']['o']    = 100;      // int, wm opacity 0 (transparent) to 100 (opaque)
 
$config['media']['video']['width']       = 300;      // px video player width
$config['media']['video']['height']      = 250;      // px video player height
$config['media']['video']['thumbWidth']  = 130;      // px thumb height
$config['media']['video']['thumbHeight'] = 85;       // px thumb height
$config['media']['video']['player'] 	 = "flowplayer.swf"; // flash player file (it has to stay in webroot/swf directory)

$config['media']['audio']['width']       = 300;      // px - audio player width
$config['media']['audio']['height']      = 24;       // px - audio player height
$config['media']['audio']['player'] 	 = "flowplayer.swf"; // flash player file (it has to stay in webroot/swf directory)


/**
 ** ******************************************
 **  Login (backend) and Security Policies
 ** ******************************************
 */

$config['maxLoginAttempts']     = 10;
$config['maxNumDaysInactivity'] = 180;
$config['maxNumDaysValidity']   = 60;


/**
 ** ******************************************
 **  More specific settings
 ** ******************************************
 */

/**
 ** Import PHP constants for smarty templates 
 */
$config['DS']        = DS;

/**
 * Modules/objects permissions
 */
define("BEDITA_PERMS_READ",	0x1) ; // read-only module permission
define("BEDITA_PERMS_MODIFY",	0x2) ;
define("BEDITA_PERMS_READ_MODIFY",	BEDITA_PERMS_READ|BEDITA_PERMS_MODIFY) ; // read-write module permission
define("OBJ_PERMS_READ_FRONT",	0x1) ; // frontend access permission
define("OBJ_PERMS_WRITE",		0x2) ; // write permission on object

/**
 * Permission params, for Smarty 
 */
$config['BEDITA_PERMS_READ'] 	= BEDITA_PERMS_READ ;
$config['BEDITA_PERMS_MODIFY'] 	= BEDITA_PERMS_MODIFY ;
$config['BEDITA_PERMS_READ_MODIFY'] = BEDITA_PERMS_READ_MODIFY;
$config['OBJ_PERMS_READ_FRONT'] 	= OBJ_PERMS_READ_FRONT ;
$config['OBJ_PERMS_WRITE'] 	= OBJ_PERMS_WRITE ;


/**
 * BEdita modules
 */
$config['modules'] = array(
	"areas" => array("id" => 1, "label" => "publishing"),
	"admin" => array("id" => 2, "label" => "admin"),
	"translations" => array("id" => 3, "label" => "translations"),
	"documents" => array("id" => 6, "label" => "documents"),
	"news" => array("id" => 7, "label" => "news"),
	"galleries" => array("id" => 8, "label" => "galleries"),
	"events" => array("id" => 10, "label" => "events"),
	"bibliographies" => array("id" => 11, "label" => "bibliographies"),
	"webmarks" => array("id" => 12, "label" => "webmarks"),
	"books" => array("id" => 13, "label" => "books"),
	"questionnaires" => array("id" => 14, "label" => "questionnaires"),
	"addressbook" => array("id" => 16, "label" => "addressbook"),
	"newsletter" => array("id" => 18, "label" => "newsletter"),
	"statistics" => array("id" => 23, "label" => "statistics"),
	"tags" => array("id" => 24, "label" => "tags"),
	"comments" => array("id" => 25, "label" => "comments"),
	"multimedia" => array("id" => 26, "label" => "multimedia")
);

/**
 * object types, main properties
 */
$config['objectTypes'] = array(
	1			=> array("id" => 1, "name" => "area", "module" => "areas", "model" => "Area"),
	"area"		=> array("id" => 1, "name" => "area", "module" => "areas", "model" => "Area"),
	
	3			=> array("id" => 3, "name" => "section", "module" => "areas", "model" => "Section"),
	"section"	=> array("id" => 3, "name" => "section", "module" => "areas", "model" => "Section"),
	
	10			=> array("id" => 10, "name" => "befile", "module" => "multimedia", "model" => "BEFile"),
	"befile"	=> array("id" => 10, "name" => "befile", "module" => "multimedia", "model" => "BEFile"),
	
	12			=> array("id" => 12, "name" => "image", "module" => "multimedia", "model" => "Image"),
	"image"		=> array("id" => 12, "name" => "image", "module" => "multimedia", "model" => "Image"),
	
	13			=> array("id" => 13, "name" => "comment", "module" => "comments", "model" => "Comment"),
	"comment"	=> array("id" => 13, "name" => "comment", "module" => "comments", "model" => "Comment"),
	
	18			=> array("id" => 18, "name" => "shortnews", "module" => "news", "model" => "ShortNews"),
	"shortnews"	=> array("id" => 18, "name" => "shortnews", "module" => "news", "model" => "ShortNews"),
	
	19			   => array("id" => 19, "name" => "bibliography", "module" => "bibliographies", "model" => "Bibliography"),
	"bibliography" => array("id" => 19, "name" => "bibliography", "module" => "bibliographies", "model" => "Bibliography"),
	
	20			=> array("id" => 20, "name" => "book", "module" => "books", "model" => "Book"),
	"book"		=> array("id" => 20, "name" => "book", "module" => "books", "model" => "Book"),
	
	21			=> array("id" => 21, "name" => "event", "module" => "events", "model" => "Event"),
	"event"		=> array("id" => 21, "name" => "event", "module" => "events", "model" => "Event"),
	
	22			=> array("id" => 22, "name" => "document", "module" => "documents", "model" => "Document"),
	"document"	=> array("id" => 22, "name" => "document", "module" => "documents", "model" => "Document"),
	
	29			=> array("id" => 29, "name" => "gallery", "module" => "galleries", "model" => "Gallery"),
	"gallery"	=> array("id" => 29, "name" => "gallery", "module" => "galleries", "model" => "Gallery"),
	
	30			  => array("id" => 30, "name" => "application", "module" => "multimedia", "model" => "Application"),
	"application" => array("id" => 30, "name" => "application", "module" => "multimedia", "model" => "Application"),
	
	31			=> array("id" => 31, "name" => "audio", "module" => "multimedia", "model" => "Audio"),
	"audio"		=> array("id" => 31, "name" => "audio", "module" => "multimedia", "model" => "Audio"),
	
	32			=> array("id" => 32, "name" => "video", "module" => "multimedia", "model" => "Video"),
	"video"		=> array("id" => 32, "name" => "video", "module" => "multimedia", "model" => "Video"),
	
	33			=> array("id" => 33, "name" => "link", "module" => "webmarks", "model" => "Link"),
	"link"		=> array("id" => 33, "name" => "link", "module" => "webmarks", "model" => "Link"),
	
	34			=> array("id" => 34, "name" => "card", "module" => "addressbook", "model" => "Card"),
	"card"		=> array("id" => 34, "name" => "card", "module" => "addressbook", "model" => "Card"),
	
	35			  => array("id" => 35, "name" => "mailmessage", "module" => "newsletter", "model" => "MailMessage"),
	"mailmessage" => array("id" => 35, "name" => "mailmessage", "module" => "newsletter", "model" => "MailMessage"),
	
	36			   => array("id" => 36, "name" => "mailtemplate", "module" => "newsletter", "model" => "MailTemplate"),
	"mailtemplate" => array("id" => 36, "name" => "mailtemplate", "module" => "newsletter", "model" => "MailTemplate"),
	
	37			 => array("id" => 37, "name" => "author", "module" => null, "model" => "Author"),
	"author" 	 => array("id" => 37, "name" => "author", "module" => null, "model" => "Author"),
	
	38			 => array("id" => 38, "name" => "biblioitem", "module" => null, "model" => "BiblioItem"),
	"biblioitem" 	 => array("id" => 38, "name" => "biblioitem", "module" => null, "model" => "BiblioItem"),
	
	39			 => array("id" => 39, "name" => "editornote", "module" => null, "model" => "EditorNote"),
	"editornote" 	 => array("id" => 39, "name" => "editornote", "module" => null, "model" => "EditorNote"),
	
	40			 	=> array("id" => 40, "name" => "question", "module" => "questionnaires", "model" => "Question"),
	"question" 		=> array("id" => 40, "name" => "question", "module" => "questionnaires", "model" => "Question"),
	
	41			 	=> array("id" => 41, "name" => "questionnaire", "module" => "questionnaires", "model" => "Questionnaire"),
	"questionnaire" => array("id" => 41, "name" => "questionnaire", "module" => "questionnaires", "model" => "Questionnaire"),
	
	42			 	=> array("id" => 42, "name" => "questionnaireresult", "module" => "questionnaires", "model" => "QuestionnaireResult"),
	"questionnaireresult" => array("id" => 42, "name" => "questionnaireresult", "module" => "questionnaires", "model" => "QuestionnaireResult"),
	
	// define array of objects that can be related to other
	'related'	=> array("id" => array(18,19,20,21,22,29,34,41,37,38)),

	// define array of objects that are leafs of the tree
	'leafs'		=> array("id" => array(18,19,20,21,22,29,33,34,37,38,41)),
	
	// define array of multimedia objects
	'multimedia' => array("id" => array(10,12,30,31,32))
);


/*
 * Relations - default objects' relation types
 * 
 * Array("defaultObjRelationType" => 
 * 		array(
 * 			"hidden" => define if relation is hidden. Used in object's view in relationship tab,
 * 			"left" => array of object_type_id related to right object_type_id (empty array means all object_type_id) 
 * 			"right => array of object_type_id related to left object_type_id (empty array means all object_type_id)
 * 			
 * OPTIONAL "objectType" => array of object_type_id related to objectType. If defined it's used instead of 'left', 'right'
 * 		)
 *	)
 * 
 */
$config["defaultObjRelationType"] = array(

	"language" => array(
		"hidden" => true,
		"left" 		=> array(),
		"right" 		=> array()	
	),
	"seealso" => array(
		"hidden" => false,
		"left" 		=> array(18,19,20,21,22,29,34),
		"right" 		=> array()	
	),
	"download" => array(
		"hidden" => false,
		"left" 		=> array(10,12,30,31,32),
		"right" 		=> array()	
	),
	"gallery" => array(
		"hidden" => true,
		"left" 		=> array(29),
		"right" 		=> array()	
	),
	"attach" => array(
		"hidden" => true,
		"left" => array(10,12,30,31,32),
		"right" => array()
	),
	"link" => array(
		"hidden" => true,
		"left" => array(33),
		"right" => array()
	),
	"author" => array(
		"hidden" => true,
		"left" => array(20),
		"right" => array(37)
	),
	"bibliography" => array(
		"hidden" => true,
		"left" => array(19),
		"right" => array(38)
	),
	"question" => array(
		"hidden" => true,
		"left" => array(41),
		"right" => array(40)
	)
);

// Relations - local objects' relation types (override in bedita.cfg)
$config["objRelationType"] = array ();

/**
 * One-way relation
 * array of relations' label that define one-way relations
 */ 
$config["defaultOneWayRelation"] = array();

// Cfg One-way relation (override in bedita.cfg)
$config["cfgOneWayRelation"] = array();

// Default reserved words [avoided in nickname creation]
$config["defaultReservedWords"] = array("section", "content", "rss", "feed", 
	"download", "xml", "xmlobject", "sitemap", "sitemap.xml", "sitemap_xml", 
	"json", "captchaImage", "saveComment", "search", "tag", "login", "logout", 
	"hashjob", "subscribe", "printme");

// Cfg reserved words (override in bedita.cfg)
$config["cfgReservedWords"] = array();

// download - redirect extensions to mediaURL [FrontenController::download]
$config["redirectExtensionsDownload"] = array ("gz", "tar", "zip");
// download - redirect mimetypes to mediaURL [FrontenController::download]
$config["redirectMimeTypesDownload"] = array ();

/**
 * Session handling parameters
 */
$config['session'] = array (
	"sessionUserKey"	=> "BEAuthUser", // session var name of info user connected
) ;


/**
 * question types
 */
$config['questionTypes'] = array(
	"multiple" => "multiple choiche",
	"single_radio" => "radio single choice",
	"single_pulldown" => "pulldown single choice",
	"freetext" => "open answer",
	"checkopen" => "check open",
	"degree" => "degree"
);

/**
 * question difficulty 
 */
$config['questionDifficulty'] = array(
	1 => "very easy",
	2 => "easy",
	3 => "normal",
	4 => "hard",
	5 => "very hard"
);


/**
 * education Level 
 * TODO: i18n!! how??
 */
$config['eduLevel'] = array(
	1 => "scuola primaria",
	2 => "scuola secondaria di primo grado",
	3 => "scuola secondaria di secondo grado",
	4 => "università"
);

/**
 * 
 * Options status select
 * 
 */
$config['statusOptions'] = array(
	"on"	=> "ON",
	"off"	=> "OFF",
	"draft"	=> "DRAFT"
) ;

/**
 * Lang selection options ISO-639-3 - Default language options for contents
 */
$config['langOptions'] = array(
		"ara" => "Arabic",
		"bul" => "Bulgarian",
		"cat" => "Catalan",
		"zho" => "Chinese",
		"hrv" => "Croatian",
		"ces" => "Czech",
		"dan" => "Danish",
		"nld" => "Dutch",
		"eng" => "English",
		"fin" => "Finnish",
		"fra" => "French",
		"deu" => "German",
		"ell" => "Greek",
		"heb" => "Hebrew",
		"hin" => "Hindi",
		"ita" => "Italian",
		"jpn" => "Japanese",
		"kor" => "Korean",
		"lav" => "Latvian",
		"lit" => "Lithuanian",
		"nor" => "Norwegian",
		"pol" => "Polish",
		"por" => "Portuguese",
		"ron" => "Romanian",
		"rus" => "Russian",
		"srp" => "Serbian",
		"slk" => "Slovak",
		"slv" => "Slovenian",
		"spa" => "Spanish",
		"swe" => "Swedish",
		"ukr" => "Ukrainian",
		"vie" => "Vietnamese"
) ;

// ISO-639-3 codes - User interface language options (backend)
$config['langsSystem'] = array(
	"eng"	=> "english",
	"ita"	=> "italiano",
//	"deu"	=> "deutsch",
//	"por"	=> "portuguěs"

) ;

// maps ISO-639-2 known codes to ISO-639-3
$config['langsSystemMap'] = array(
	"it"	=> "ita",
	"en"	=> "eng",
	"en_us"	=> "eng",
	"en_gb"	=> "eng",
	"es"	=> "spa",
	"pt"	=> "por",
	"pt_br"	=> "por"
) ;

// 'langsIso' empty, overridden if 'langOptionsIso' is true
$config['langsIso'] = array();
// add langs.iso.php to language options for content 
$config['langOptionsIso'] = false;

// media types for objects in multimedia module
$config['mediaTypes'] = array('image','video','audio','text','spreadsheet','presentation','drawing','chart','formula','application');


/**
 * variables used for accepting remote URL and for identifying a BEdita object type from mime  
 */
$config['validate_resource'] = array(
	'paranoid'	=> true,	// if true and 'allow_url_fopen'=false doesn't accept remte URL

	// generic URL
	'URL'	=> '/^\s*[a-z\d\+\-\.]+\:\/\//i',
	
	 // URL allowed
	'allow'	=> array( 
				'/^\s*http:\/\/(.*)\.(html|htm)$/',
				'/(ftp|http|https):\/\/(\w+:{0,1}\w*@)?(\S+)(:[0-9]+)?(\/|\/([\w#!:.?+=&%@!\-\/]))?/'
			   ),
	/**
	 * Accepted MIME types for different object types
	 * The structure is:
	 * 		ModelName => array(mime type regular expressions) 
	 * or
	 * 		ModelName => array(
	 * 			"specific type" => array(
	 * 				"mime_type" => array(mime type regular expressions),
	 * 				other parameters...
	 * 			)
	 * 		)
	 */
	'mime'	=> array(
				'Image'			=> array('/image\/\.*/'),
				'Audio'			=> array('/audio\/\.*/'),
				'Video'			=> array('/video\/\.*/','/application\/flash-video/'),
			    'Application'	=> array(
							    		"flash" => array(
												"mime_type" => array('/application\/x-shockwave-flash/'),
												"application_type" => "application/x-shockwave-flash",
												"label" => "Adobe Flash"
											)
									),
				'BEFile'		=> array('/application\/\.*/', '/text\/\.*/', '/beexternalsource/')
				)
) ;


/**
 *  Supported media providers
 *  
 *  for any providers supported are defined an array of regexp to identify provider and media id 
 *  and an array of params
 */
$config['media_providers'] = array(
	"youtube"	=> array(
		"regexp" => array(
			'/^http:\/\/\w{3}\.youtube\.com\/watch\?v=(.[^&]+)/',
			'/^http:\/\/youtube\.com\/watch\?v=(.[^&]+)/',
			'/^http:\/\/[a-z]{2}\.youtube\.com\/watch\?v=(.[^&]+)/'
		),
		"params" => array(
			"width" 	=> 300,
			"height" 	=> 250,
			"urlthumb"	=> "http://i.ytimg.com/vi/%s/default.jpg",
			"urlembed"	=> "http://www.youtube.com/oembed?url=%s"
		)
	),
	"blip"	=> array(
		"regexp" => array(
			'/^http:\/\/\w{3}\.blip\.tv\/file\/(\d+)\?{0,1}.*/',
			'/^http:\/\/blip\.tv\/file\/(\d+)\?{0,1}.*/'
		),
		"params" => array(
			"width"		=> 300,
			"height"	=> 250,
			"urlinfo" 	=> "http://www.blip.tv/file/%s?skin=json",
			"urlembed"	=> "http://blip.tv/oembed/?url=%s"
		)
	),
	"vimeo"	=> array(
		"regexp" => array(
			'/^http:\/\/\w{3}\.vimeo\.com\/(\d+)/',
			'/^http:\/\/vimeo\.com\/(\d+)/'
		),
		"params" => array(
			"width"		=> 300,
			"height"	=> 250,
			"urlinfo" 	=> "http://vimeo.com/api/v2/video/%s.%s",
			"urlembed"	=> "http://vimeo.com/api/oembed.json?url=%s"
		)
	)
) ;

/**
 *  Default model bindings for Containable Behaviour 
 */
$config['modelBindings'] = array() ;


// default email fields -- override in bedita.cfg or bedita.sys
$config['mailOptions'] = array(
	"sender" => "noreply@bedita.com", 
	"reply_to" => "noreply@bedita.com", 
	"signature" => "powered by BEdita - http://www.bedita.com",
);


/**
 *  default values for fulltext search. Override in bedita.cfg  
 */
$config['searchFields'] = array() ;

/**
 * Default css filename for newsletter templates
 */
$config['newsletterCss'] = "base.css";

/**
 * default timeout in minutes to assume mail jobs blocked and try to resend
 */
$config['newsletterTimeout'] = 15;

/**
 * Default value in milliseconds, between autosave of objects (status draft or off)
 */
$config['autosaveTime'] = 120000; // two minutes

/**
 * Hash jobs expired time in seconds (7 days)
 */
$config['hashExpiredTime'] = 604800;

/**
 * Specific System settings
 * ALWAYS AT THE END of bedita.ini.php
 */
$appPath = (defined("BEDITA_CORE_PATH"))? BEDITA_CORE_PATH . DS : APP;
if (file_exists ($appPath."config".DS."bedita.sys.php") ) {
	include_once($appPath."config".DS."bedita.sys.php") ;	
}
unset($appPath);
?>